Applications/Spire/Source/Ui/CalendarWidget.cpp

```diff
@@ -0,0 +1,145 @@
+  });
+  layout->addWidget(m_month_spin_box);
+  layout->addSpacing(scale_height(4));
+  auto day_label_layout = new QHBoxLayout();
+  day_label_layout->setSpacing(scale_width(3));
+  day_label_layout->setContentsMargins({});
+  add_day_label(day_label_layout, tr("S"));
+  add_day_label(day_label_layout, tr("M"));
+  add_day_label(day_label_layout, tr("T"));
+  add_day_label(day_label_layout, tr("W"));
+  add_day_label(day_label_layout, tr("T"));
+  add_day_label(day_label_layout, tr("F"));
+  add_day_label(day_label_layout, tr("S"));
+  layout->addLayout(day_label_layout);
+  layout->addSpacing(scale_height(3));
+  m_calendar_layout = new QGridLayout();
+  m_calendar_layout->setContentsMargins({});
+  m_calendar_layout->setSpacing(0);
+  m_calendar_layout->setHorizontalSpacing(scale_width(3));
+  m_calendar_layout->setVerticalSpacing(scale_height(3));
+  layout->addLayout(m_calendar_layout);
+  set_date(selected_date);
+}
+
+void CalendarWidget::set_date(const date& date) {
+  if(date != m_selected_date) {
+    m_calendar_model.set_month(date.month(), date.year());
+    update_calendar(date);
+  }
+  m_selected_date = date;
+  set_highlight();
+}
+
+connection CalendarWidget::connect_date_signal(
+    const DateSignal::slot_type& slot) const {
+  return m_date_signal.connect(slot);
+}
+
+void CalendarWidget::on_date_selected(const date& date) {
+  if(date != m_selected_date) {
+    if(date.month() != m_month_spin_box->get_date().month() &&
+        date.year() != m_month_spin_box->get_date().month()) {
+      update_calendar(date);
+      m_month_spin_box->set_date(date);
+    }
+    m_date_signal(date);
+  }
+  m_selected_date = date;
+  set_highlight();
+}
+
+void CalendarWidget::on_month_changed(const date& date) {
+  update_calendar(date);
+  if(std::tie(date.year(), date.month()) ==
+      std::tie(m_selected_date.year(), m_selected_date.month())) {
+    set_highlight();
+  }
+}
+
+void CalendarWidget::add_day_label(QLayout* layout, const QString& text) {
+  auto label = new QLabel(text, this);
+  label->setFixedSize(scale(20, 20));
+  label->setAlignment(Qt::AlignCenter);
+  label->setStyleSheet(QString(R"(
+      background-color: #FFFFFF;
+      color: #4B23A0;
+      font-family: Roboto;
+      font-size: %1px;
+    )").arg(scale_height(12)));
+  layout->addWidget(label);
+}
+
+CalendarDayWidget* CalendarWidget::get_day_widget(
+    const date& date) {
+  auto [x, y] = m_calendar_model.get_pos(date);
+  return static_cast<CalendarDayWidget*>(m_calendar_layout->itemAtPosition(
+    y, x)->widget());
+}
+
+void CalendarWidget::set_highlight() {
+  if(m_selected_date_widget) {
+    m_selected_date_widget->remove_highlight();
+  }
+  m_selected_date_widget = get_day_widget(m_selected_date);
+  m_selected_date_widget->set_highlight();
+}
+
+void CalendarWidget::update_calendar(const date& displayed_date) {
+  m_calendar_model.set_month(displayed_date.month(), displayed_date.year());
+  if(m_selected_date_widget) {
```

Comment:
You can remove the if guard. No need to check if it's ```nullptr``` to assign to ```nullptr``` and performing the check doesn't gain anything.